addons/account/models/account_journal_dashboard.py

```diff
@@ -117,324 +138,454 @@ def _graph_title_and_key(self):
             company = self.env['res.company'].browse(result.get('company_id')) or self.env.company
-            rslt_count += 1
             date = result.get('invoice_date') or fields.Date.context_today(self)
 
-            amount = result.get('amount_total', 0) or 0
-            if cur != target_currency:
-                key = (cur, target_currency, company, date)
-                # Using setdefault will call _get_conversion_rate, so we explicitly check the
-                # existence of the key in the cache instead.
+            if document_currency == target_currency:
+                total_amount += result.get('amount_total') or 0
+            elif company.currency_id == target_currency:
+                total_amount += result.get('amount_total_company') or 0
+            else:
+                key = (document_currency, target_currency, company, date)
                 if key not in curr_cache:
                     curr_cache[key] = self.env['res.currency']._get_conversion_rate(*key)
-                amount *= curr_cache[key]
-            rslt_sum += target_currency.round(amount)
-        return (rslt_count, rslt_sum)
+                total_amount += (result.get('amount_total') or 0) * curr_cache[key]
+        return (len(results_dict), target_currency.round(total_amount))
+
+    def _get_journal_dashboard_bank_running_balance(self):
+        # In order to not recompute everything from the start, we take the last
+        # bank statement and only sum starting from there.
+        self._cr.execute("""
+            SELECT journal.id AS journal_id,
+                   statement.id AS statement_id,
+                   COALESCE(statement.balance_end_real, 0) AS balance_end_real,
+                   without_statement.amount AS unlinked_amount,
+                   without_statement.count AS unlinked_count
+              FROM account_journal journal
+         LEFT JOIN LATERAL (  -- select latest statement based on the date
+                           SELECT id,
+                                  date,
+                                  balance_end_real
+                             FROM account_bank_statement
+                            WHERE journal_id = journal.id
+                         ORDER BY date DESC
+                            LIMIT 1
+                   ) statement ON TRUE
+         LEFT JOIN LATERAL (  -- sum all the lines not linked to a statement with a higher index than the last line of the statement
+                           SELECT COALESCE(SUM(stl.amount), 0.0) AS amount,
+                                  COUNT(*)
+                             FROM account_bank_statement_line stl
+                             JOIN account_move move ON move.id = stl.move_id
+                            WHERE stl.statement_id IS NULL
+                              AND move.state != 'cancel'
+                              AND move.journal_id = journal.id
+                              AND move.date >= statement.date
```

Comment:
should be using internal_index for the clause, because the proposed code would fail if a transaction has no statement but is on the same date as a bank statement's last (and only) line but shouldn't be included (because prior to the statement)

Reply:
Fixed